various fixups for atomic operation APIs

tyrantlizard77 opened the pull request:
- fix return value type of atomic_cas() being bool in some cases, int in others
- add atomic_ptr_t typedef and atomic_ptr_cas(), atomic_ptr_set(), atomic_ptr_get(), atomic_ptr_clear() APIs
- fix a type safety issue in shell

Fixes: #22887

Commit:
kernel: fix return type for atomic_cas()

In some cases this was a bool, in some cases an integer value
of 1 or 0. Standardize on bool.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Requires a Nordic HAL patch:

```
From c4ad92e9347c18c1312263055587c2f9554da194 Mon Sep 17 00:00:00 2001
From: Andrew Boie <andrew.p.boie@intel.com>
Date: Tue, 18 Feb 2020 12:02:28 -0800
Subject: [PATCH] nrfx_usbd: fix data type for atomic variable
m_ep_dma_waiting is passed to various NRFX_ATOMIC functions,
but is defined as uint32_t. Depending on the implementation,
this may cause warnings if this is not the same data type as
nrfx_atomic_t. Change this to match.
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
---
 nrfx/drivers/src/nrfx_usbd.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
diff --git a/nrfx/drivers/src/nrfx_usbd.c b/nrfx/drivers/src/nrfx_usbd.c
index c0598e0..0f6d67b 100644
--- a/nrfx/drivers/src/nrfx_usbd.c
+++ b/nrfx/drivers/src/nrfx_usbd.c
@@ -269,7 +269,7 @@ static uint32_t m_ep_ready;
  * Mask prepared USBD data for transmission.
  * It is cleared when no more data to transmit left.
  */
-static uint32_t m_ep_dma_waiting;
+static nrfx_atomic_t m_ep_dma_waiting;
 /**
  * @brief Current EasyDMA state.
-- 
2.25.0
```

nordic-krch commented:
@andrewboie given your comment (https://github.com/zephyrproject-rtos/zephyr/issues/22887#issuecomment-587590041) should then `atomic_t` become `uintptr_t` or `slong_t`? What's pros and cons? They are same size on 32bit and 64bit so maybe it does not matter much.

nordic-krch commented:
I think you can add "fixes #22887" to the description.

pfalcon commented on tests/kernel/common/src/atomic.c:
```diff
@@ -12,6 +12,16 @@
+#define INITIAL		0x0f0f0f0f0f0f0f0fL
+#define NAND_RESULT	0xFFFFFFFFFFFFF0FFL
+#else
+#define MAXBIT		32
```
Did you consider giving this different name, like NUMBITS. Because I'd consider for 32 bits, "max bit" to be 31.

Commit:
shell: use correct data type for atomic variables

The variable needs to be an atomic_t, and in one case it was
being incremented outside of an atomic_inc/atomic_add.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
> @andrewboie given your comment ([#22887 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/22887#issuecomment-587590041)) should then `atomic_t` become `uintptr_t` or `slong_t`? What's pros and cons? They are same size on 32bit and 64bit so maybe it does not matter much.

atomic_t is currently a signed integer. using uintptr_t would convert it to an unsigned value, which I do not want to do in this PR.

joerchan commented:
> > @andrewboie given your comment ([#22887 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/22887#issuecomment-587590041)) should then `atomic_t` become `uintptr_t` or `slong_t`? What's pros and cons? They are same size on 32bit and 64bit so maybe it does not matter much.
> 
> atomic_t is currently a signed integer. using uintptr_t would convert it to an unsigned value, which I do not want to do in this PR.

Is a long guaranteed to be the same size as a pointer? AFAIK it is only guaranteed to be 32 bits or more.

andrewboie commented:
> > > @andrewboie given your comment ([#22887 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/22887#issuecomment-587590041)) should then `atomic_t` become `uintptr_t` or `slong_t`? What's pros and cons? They are same size on 32bit and 64bit so maybe it does not matter much.
> > 
> > 
> > atomic_t is currently a signed integer. using uintptr_t would convert it to an unsigned value, which I do not want to do in this PR.
> 
> Is a long guaranteed to be the same size as a pointer? AFAIK it is only guaranteed to be 32 bits or more.

I used one of zephyr's derived types in zephyr/types.h. I did not typedef atomic_t to long.

slong_t / ulong_t are defined as being able to hold a pointer. They are that header's version of uintptr_t / intptr_t

andyross commented:
We can't make atomic_t be an architecture-dependent size.  Users need to be able to store known quantities in it.   And on a practical level: 64 bit platforms need to be able to make atomic updates to 32 bit ints too.

andrewboie commented:
> We can't make atomic_t be an architecture-dependent size. Users need to be able to store known quantities in it. And on a practical level: 64 bit platforms need to be able to make atomic updates to 32 bit ints too.

What do you suggest we do?
I'm loath to introduce a parallel set of 64-bit wide atomic APIs, signed vs unsigned, etc

andyross commented:
I don't see that we have much choice.  Apps want a consistent data type for atomic operations (e.g lock counts) that isn't architecture-dependent.  But they also want to be able to store a pointer/address in them.  I think having a separate "atomic_addr_t" (which on 32 bit systems would just be implemented as a atomic_t) is reasonable enough.

And I take no position on signed/unsigned.  From my perspective it's perfectly OK to have it be any integer type and let the applications cast as needed.  The feature being provided is about the memory underneath the type, not the bytes themselves.

joerchan commented:
What about the original suggestion:
> Create atomic_ptr_t type and set of atomic_ptr_cas(), atomic_ptr_set(), atomic_ptr_get() and atomic_ptr_clear() functions.

andrewboie commented:
> What about the original suggestion:
> 
> > Create atomic_ptr_t type and set of atomic_ptr_cas(), atomic_ptr_set(), atomic_ptr_get() and atomic_ptr_clear() functions.

Yeah, you guys have convinced me ... updated PR in a bit

Commit:
kernel: add APIs for atomic os on pointers

The existing APIs are insufficient on 64-bit systems as atomic_t
is 32-bits wide.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Updated PR ... adds atomic_ptr_* APIs for cas, get, set, clear.
